### PR TITLE
Support distributed encoding

### LINKED ISSUE #10: Support distributed encoding
This has been a highly requested feature. 

Process:

1. Run analysis on primary node.
1. For each segment:
    1. Wait for available worker node.
    1. Load segment video data on primary node.
    1. Send video data to worker node (xz or similarly compressed).
    1. Have some protocol to communicate progress data back and forth.
    1. Receive completed encode back from worker node.
1. After all segments completed, merge them together on primary node.

### PR DESCRIPTION
**Pull request by shssoichiro**
Adds rav1e-worker, which acts as a server on a remote encoding machine and listens to connections from clients. The client, part of rav1e-by-gop, negotiates with the server, the server notifies the client as it has slots available, and the client sends one segment at a time to the server. The server can have multiple worker threads, and a client can use as many threads as the server will allow. A server can also serve multiple clients at once, which it does via a FIFO queue. The queue only holds slot requests, so the client does not send frames until the server signals a slot is ready. The client will send to the first worker that signals a slot is available. This keeps memory usage down on both the client and server, and streamlines the threading process so the end of an encode is not stalled waiting for a specific remote worker, if there are other workers or local threads available.

Communication is done using websockets, and is insecure by default, but can be configured to use TLS. Unfortunately this does not currently work with self-signed certs. Client-server authentication is done using a simple password system; a password is configured on the server, and then the client must be configured to use the same password for this server.

The frame sending process is quite upload-heavy. The input frames are sent xz-compressed to try and alleviate this, however results will still be best if workers are on a LAN or other connection with a fast upload. I found that even with a 10Mbps upload from my home connection, the remote workers still were more than efficient enough to be worth using.

Closes #10

### CODE REVIEW

**Comment by lu-zero** [comment]
Can we split this PR in 3?
- Refactor (it should land immediately)
- Client/Server structure (should enjoy some patch juggling)
- Refinements (it might require some discussion and I might have few more suggestions)
  - raw content compression (e.g. xz or zstd?)
  - auth

**Comment by shssoichiro** [comment]
@lu-zero I'm opting to reuse this PR for discussion; I've rebased off master. I think it's fine to have the discussion on compression and auth here as well. If we need to split something out into another PR e.g. more complex auth support, that's reasonable, but at the moment the compression and auth implementations are fairly small code-wise.

**Comment by shssoichiro** [comment]
Seems to be some big memory leaking going on... so, going to investigate.

File: rav1e-by-gop/src/bin/encode.rs [hunk 1]
```diff
@@ -138,19 +185,96 @@ pub fn perform_encode_inner<T: Pixel, R: 'static + Read + Send>(
...
             known_keyframes,
             skipped_segments,
-        )
-        .expect("An error occurred during input analysis");
+            video_info,
+            s,
+        );
     });
 
+    for worker in remote_slots.lock().unwrap().iter() {
```
Comment (lu-zero) [inline rav1e-by-gop/src/bin/encode.rs:222]: This block could be a stand alone function maybe?

File: rav1e-by-gop/src/bin/encode.rs [hunk 2]
```diff
@@ -166,17 +290,53 @@ pub fn perform_encode_inner<T: Pixel, R: 'static + Read + Send>(
     let mut num_segments = 0;
     let encode_opts = EncodeOptions::from(opts);
     let output_file = opts.output.to_owned();
+
```
Comment (lu-zero) [inline rav1e-by-gop/src/bin/encode.rs]: this as well.

File: rav1e-worker/src/worker.rs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            let _ = connection.socket.write_message(Message::Binary(
+                if slot_request.video_info.bit_depth <= 8 {
+                    rmp_serde::to_vec(&EncoderMessage::SlotAllocated::<u8>(connection.id)).unwrap()
+                } else {
+                    rmp_serde::to_vec(&EncoderMessage::SlotAllocated::<u16>(connection.id)).unwrap()
+                },
+            ));
+
+            // Spawn the encoder thread for this connection within the encoder thread pool
+            thread_pool.execute(move || {
```
Comment (lu-zero) [inline rav1e-worker/src/worker.rs]: This should be a stand alone function as well.

File: rav1e-worker/src/worker.rs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+pub fn encode_segment<T: Pixel + Default + Serialize>(
+    mut connection: ConnectedSocket,
+    encode_request: SlotRequestMessage,
+    input: RawFrameData<T>,
+) {
+    let connection_id = input.connection_id;
+    let frames = input.frames.into_iter().map(Arc::new).collect::<Vec<_>>();
+    if !frames.is_empty() {
+        let mut cfg = Config {
```
Comment (lu-zero) [inline rav1e-worker/src/worker.rs]: This could be a stand alone function as well.

**Review by lu-zero — COMMENTED** [review]
I managed to read it again and I put few initial comments.

Where is the buildup/leak happening?

**Comment by shssoichiro** [comment]
I'm only noticing the high memory usage when dealing with 1080p+ videos and a slow upload to the remote server (slow = 10mbps). It may be the leak is still occurring in other situations but is not as noticeable. I haven't been able to track down what's causing it, but my suspicions are either xz or the websocket connection or underlying network connection itself is causing it.

**Comment by lu-zero** [comment]
I guess frames end up piling up somewhere.

**Comment by shssoichiro** [comment]
I'm going to do some more testing on the memory usage and see if I can pin it down. In the meantime, @lu-zero you mentioned you had some opinions on compression and authorization. Would you mind sharing those?

**Comment by lu-zero** [comment]
Sure:)
- Regarding compression probably zstd or lz4 might have a better tradeoff compared to xz.
- For auth I hadn't read in depth what's in place. 
  - I guess some deploys would enjoy having no-auth while other might prefer having other auth systems.
- Transport-wise tls/plain is good.

**Comment by shssoichiro** [comment]
Fixed a couple of bugs, and switched over to zstd. Zstd does seem to provide a better speed-to-compression ratio. I can't even seem to replicate the memory issue consistently, so for all I know it might be KDE doing some weird thumbnailing stuff or something.

The current auth implementation uses a simple password shared between the server and client. The client sends the password as a header when it makes the connection, and the server rejects the connection if the password doesn't match. A password is currently required--I really don't like having no-password as a default, so maybe there could be an explicit no-password option if that's a desire for some users.

**Comment by lu-zero** [comment]
Since the tls transport could mutually identify the endpoints a password would be unnecessary :) but those are small details I guess.

**Comment by lu-zero** [comment]
Since the tls transport could mutually identify the endpoints a password would be unnecessary :) but those are small details I guess.

**Comment by shssoichiro** [comment]
In this case I'm not sure how TLS would be used for what I'm trying to accomplish with the password. Since TLS automatically assumes the client is trustworthy if it has the required ciphers installed. The intent behind the password is, for example, if I have a worker running on a server that's internet-facing, I want to prevent random users who may find out that the port is open from loading up my worker with tasks and CPU usage. (This is such obscure software at the moment that it's unlikely to be targeted by such malicious users, but even in a case like this, I try to avoid security through obscurity.)

**Comment by tleydxdy** [comment]
TLS can have client certificate

**Comment by shssoichiro** [comment]
After some consideration, I thought about using the TLS client cert, but I'd prefer to adopt an approach that also works for unencrypted connections and is secured by default. In reality, I'd love to make this TLS-only, but given that I was unable to get it to work with self-signed certs, that's not going to be feasible for people who want to use this over a LAN. So I think that requiring a password by default is a reasonable compromise.

**Review by lu-zero — APPROVED** [review]
I guess we can refine after :)

**Pull request merged**